jpype/_jarray.pyi

```diff
@@ -0,0 +1,54 @@
+# *****************************************************************************
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+#   See NOTICE file for details.
+#
+# *****************************************************************************
+import collections.abc
+import typing
+
+
+__all__ = ['JArray']
+
+
+T = typing.TypeVar('T')
+
+
+class JArray(collections.abc.Collection[T], collections.abc.Reversible[T]):
+
+    def __new__(cls, tp, dims=1):
+        ...
+
+    @classmethod
+    def of(cls, array, dtype=None):
+        ...
+
+    def __len__(self):
+        ...
+
+    def __iter__(self):
+        ...
+
+    def __contains__(self, item):
+        # this is implicitly implemented
+        ...
+
+    def __reversed__(self):
```

Comment:
They should already be inferred by the abstract base classes. I put the function stubs in because mypy would otherwise complain that it's not an abstract class but it's not defining those functions. I figured it was better to do this then set the meta class as abc.ABCMeta because it's not accurate.

I don't see a reason to not put in the type hints though, rather do it and ensure any type checker gets it right. Will fix it tonight.

Reply:
As strange as this may sound, I copied the signatures directly from `Python3/Lib/_collections_abc.py` to ensure I got them correct. I'm going to add them anyway though.

Reply:
Please sanity check the applied type hints

Reply:
Ok so after further investigation, I discovered that `isinstance(JInt[5], typing.Collection)` will return `False` even though it meets the criteria. I'm assuming this is because `__contains__` is implicit and unfortunately, for some reason, even if explicitly added to `JArray` it still doesn't have the attribute.

This should finally settle everything I think.